services/src/main/java/org/keycloak/services/util/DPoPUtil.java

```diff
@@ -0,0 +1,401 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.util;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.ws.rs.core.UriBuilder;
+import javax.ws.rs.core.UriInfo;
+
+import org.apache.commons.codec.binary.Hex;
+
+import org.keycloak.TokenVerifier;
+import org.keycloak.common.VerificationException;
+import org.keycloak.common.util.Time;
+import org.keycloak.crypto.Algorithm;
+import org.keycloak.crypto.KeyWrapper;
+import org.keycloak.crypto.SignatureProvider;
+import org.keycloak.crypto.SignatureVerifierContext;
+import org.keycloak.exceptions.TokenVerificationException;
+import org.keycloak.http.HttpRequest;
+import org.keycloak.jose.jwk.JWK;
+import org.keycloak.jose.jws.JWSHeader;
+import org.keycloak.jose.jws.crypto.HashUtils;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.SingleUseObjectProvider;
+import org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.dpop.DPoP;
+import org.keycloak.util.JWKSUtils;
+
+import static org.keycloak.utils.StringUtil.isNotBlank;
+
+/**
+ * @author <a href="mailto:dmitryt@backbase.com">Dmitry Telegin</a>
+ */
+public class DPoPUtil {
+
+    public static final int DEFAULT_PROOF_LIFETIME = 10;
+    public static final int DEFAULT_ALLOWED_CLOCK_SKEW = 2;
+    public static final String DPOP_TOKEN_TYPE = "DPoP";
+    public static final String DPOP_SCHEME = "DPoP";
+    public final static String DPOP_SESSION_ATTRIBUTE = "dpop";
+    public final static String DPOP_PARAM = "dpop";
+    public final static String DPOP_THUMBPRINT_NOTE = "dpop.thumbprint";
+
+    public static enum Mode {
+        ENABLED,
+        OPTIONAL,
+        DISABLED
+    }
+
+    private static final String DPOP_HTTP_HEADER = "DPoP";
+    private static final String DPOP_JWT_HEADER_TYPE = "dpop+jwt";
+    private static final String DPOP_ATH_ALG = "RS256";
+
+    public static final Set<String> DPOP_SUPPORTED_ALGS = Stream.of(
+        Algorithm.ES256,
+        Algorithm.ES384,
+        Algorithm.ES512,
+        Algorithm.PS256,
+        Algorithm.PS384,
+        Algorithm.PS512,
+        Algorithm.RS256,
+        Algorithm.RS384,
+        Algorithm.RS512
+    ).collect(Collectors.toSet());
+
+    private static URI normalize(URI uri) {
```

Comment:
It might be OK. However, the actual URI normalization seems to follow the following procedures:

[6.2.2.  Syntax-Based Normalization](
https://www.rfc-editor.org/rfc/rfc3986.html#section-6.2.2)
[6.2.3.  Scheme-Based Normalization](
https://www.rfc-editor.org/rfc/rfc3986.html#section-6.2.3)